### PR TITLE
Core: Extend ResolvingFileIO to support prefix-based operations

### PR DESCRIPTION
**Pull request by nastra**

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java [hunk 1]
```diff
@@ -252,4 +254,22 @@ protected void finalize() throws Throwable {
       }
     }
   }
+
+  @Override
+  public Iterable<FileInfo> listPrefix(String prefix) {
+    return withPrefixSupport(io(prefix)).listPrefix(prefix);
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java:260]: Technically, there could be code that checks `if (io instanceof SupportsPrefixOperations) { // do some prefix-related operation }` where `io` is of type `ResolvingFileIO` but the underlying io implementation doesn't support  prefix operations. We unfortunately don't have a good way to detect this

Comment (RussellSpitzer) [reply core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java:260]: I was worried about this on the previous PR that suggested it. We basically force all delegates of Resolving fileIO to now support prefix operations, but we don't actually say this anywhere. If we are going to make this change we should have Asserts in the init method, break the rev api, properly type "io" to return a FileIO which extends SupportsPrefix and so forth.

Comment (nastra) [reply core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java:260]: In that case we would end up pretty much with what was proposed in https://github.com/apache/iceberg/pull/8169, which is OK I guess?

Comment (bryanck) [reply core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java:260]: I think if we want to add prefix operations we'll need something like that PR, as we can't fall back to a default implementation like with bulk deletes.

File: core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java [hunk 2]
```diff
@@ -95,25 +100,12 @@ public void deleteFiles(Iterable<String> pathsToDelete) throws BulkDeletionFailu
     Iterators.partition(pathsToDelete.iterator(), BATCH_SIZE)
         .forEachRemaining(
             partitioned -> {
-              Map<FileIO, List<String>> pathByFileIO =
+              Map<DelegateFileIO, List<String>> pathByFileIO =
```
Comment (bryanck) [inline core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java:103]: We can just call the delegate directly now, I don't believe there is a need for this logic anymore

File: core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java [hunk 3]
```diff
@@ -252,4 +252,22 @@ protected void finalize() throws Throwable {
...
+  public Iterable<FileInfo> listPrefix(String prefix) {
+    return withPrefixSupport(io(prefix)).listPrefix(prefix);
+  }
+
+  @Override
+  public void deletePrefix(String prefix) {
+    withPrefixSupport(io(prefix)).deletePrefix(prefix);
+  }
+
+  private SupportsPrefixOperations withPrefixSupport(FileIO io) {
```
Comment (bryanck) [inline core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java:266]: We don't need this I don't think if we're assuming DelegateFileIO is implemented

Comment (nastra) [reply core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java:266]: oops, you're right. This was a leftover

File: core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java (hunk 2 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java:103]: do you think we should still batch here or would we want to essentially use the same implementation you had?
```
// peek at the first element to determine the type of FileIO
    Iterator<String> originalIterator = pathsToDelete.iterator();
    if (!originalIterator.hasNext()) {
      return;
    }

    PeekingIterator<String> iterator = Iterators.peekingIterator(originalIterator);
    DelegateFileIO fileIO = io(iterator.peek());
    fileIO.deleteFiles(() -> iterator);
```

To me it seems like batching should probably be handled by the underlying FileIO implementation, but just wanted to see what others think?

Comment (bryanck) [reply core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java:103]: actually right, it is probably better to keep this as there may be multiple FileIO types.

File: core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java (hunk 1 above)
Comment (RussellSpitzer) [reply core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java]: I'm fine with the approach in #8169 But we need to bite the bullet. Either we are breaking this api and if so we should make all of our requirements clear. The class has to fail if it is init with a non compatible class and we should remove all cast's.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java [hunk 4]
```diff
@@ -76,7 +77,11 @@
  * will result in {@link org.apache.iceberg.exceptions.ValidationException}.
  */
 public class S3FileIO
-    implements FileIO, SupportsBulkOperations, SupportsPrefixOperations, CredentialSupplier {
+    implements FileIO,
+        SupportsBulkOperations,
```
Comment (RussellSpitzer) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:81]: These are duplicate now right?

File: core/src/main/java/org/apache/iceberg/hadoop/HadoopFileIO.java [hunk 5]
```diff
@@ -48,7 +49,11 @@
 import org.slf4j.LoggerFactory;
 
 public class HadoopFileIO
-    implements FileIO, HadoopConfigurable, SupportsPrefixOperations, SupportsBulkOperations {
+    implements FileIO,
+        HadoopConfigurable,
```
Comment (RussellSpitzer) [inline core/src/main/java/org/apache/iceberg/hadoop/HadoopFileIO.java:53]: Same here on duplicate implementations

File: core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java [hunk 6]
```diff
@@ -31,18 +31,23 @@
...
 
-/** FileIO implementation that uses location scheme to choose the correct FileIO implementation. */
-public class ResolvingFileIO implements FileIO, HadoopConfigurable, SupportsBulkOperations {
+/**
+ * FileIO implementation that uses location scheme to choose the correct FileIO implementation.
+ * Delegate FileIO implementations must implement the {@link DelegateFileIO} mixin interface,
+ * otherwise initialization will fail.
+ */
+public class ResolvingFileIO
+    implements FileIO, HadoopConfigurable, SupportsBulkOperations, SupportsPrefixOperations {
```
Comment (RussellSpitzer) [inline core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java:50]: could just implement DelgateFileIO itself here

File: core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java [hunk 7]
```diff
@@ -213,10 +206,17 @@ FileIO io(String location) {
               }
             }
           }
+
+          Preconditions.checkState(
```
Comment (RussellSpitzer) [inline core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java:210]: This is fine, but we could also do the cast in the try above  and catch the class cast exception. I'm fine with either approach and think there are benefits to both.

File: core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java (hunk 6 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java]: fixed

File: core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java (hunk 7 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java:209]: I'd slightly prefer the check here as it's more explicit

File: core/src/main/java/org/apache/iceberg/hadoop/HadoopFileIO.java (hunk 5 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/hadoop/HadoopFileIO.java]: fixed

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java (hunk 4 above)
Comment (nastra) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java]: fixed

File: core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java (hunk 1 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java]: should be fixed now

**Comment by RussellSpitzer** [comment]
Thanks @nastra for finishing this up and @bryanck for the foundational work and review.

**Comment by leoluan2009** [comment]
@nastra @RussellSpitzer Should ADLSFileIO implement DelgateFileIO as GCSFileIO to make code clean and consistently?

**Comment by nastra** [comment]
@leoluan2009 yes it should, good catch. I've opened https://github.com/apache/iceberg/pull/8563 to address that

**Pull request merged**